Adds handling of recurring payment

kmmatwork opened the pull request:
Note that the acceptance tests for recurring payment will not run
unless the NGROK_AUTH_TOKEN env. variable has been set.

Commit:
Adds customer.subscription.updated event to list of subscribed to Stripe events

Commit:
Adds handling of Stripe events related to subscription renewal

Follows the recommodation as described at:

   https://stripe.com/docs/billing/subscriptions/payment

Commit:
Adds client API for reattemptng the renewal of a subscription that
previously failed due to failed payment

Commit:
Adds acceptance tests for recurring payment

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

vihangpatil commented on acceptance-tests/Dockerfile.ngrok:
```diff
@@ -0,0 +1,18 @@
+FROM azul/zulu-openjdk:13
```
Today, I have bumped this version to 13.0.1 which is already merged to develop.

vihangpatil commented on acceptance-tests/src/main/kotlin/org/ostelco/at/TestSuite.kt:
```diff
@@ -49,6 +51,41 @@ class TestSuite {
+     * skipped altogether.
+     */
+    @Test
+    @EnabledIfEnvironmentVariable(named = "NGROK_AUTH_TOKEN", matches="\\S+")
```
Do you want these tests to be run on CI?  Should we add this ENV var to CI setup?

kmmatwork replied:
Yes, but we will first have to create an account for CI with ngrok.io first. And maybe consider a baslc or pro subscription? Being able to use "custom subdomain" would simplify the setup a bit...

vihangpatil commented on model/src/main/kotlin/org/ostelco/prime/model/Entities.kt:
```diff
@@ -297,26 +291,26 @@ enum class ProductProperties(val s: String) {
 
 enum class ProductClass {
     SIMPLE_DATA,
+    SUBSCRIPTION,
```
Subscription is not a Product class.  Instead, it is a payment type.
(Product Class = Membership + Payment type = Subscription) => Our Membership project which we have removed.
(Product Class = Simple Data + Payment type = Subscription) => Product which will topup `x` bytes evey `y` period.
(Product Class = Data with Validity (coming soon) + Payment type = Subscription) => We will have this.

kmmatwork replied:
Maybe it could be an idea to rewrite the "payment paths" in the code to only consider the `payment[type]` field and not consider the `productClass` field at all? And maybe make the `payment[type]` field mandatory?

kmmatwork replied:
Don't know yet how straight forward this will be though. Will have to check first...

vihangpatil commented on neo4j-store/src/main/resources/AcceptanceTestSetup.kts:
```diff
@@ -126,4 +129,71 @@ adminStore.atomicCreateOffer(
+                sku = "PLAN_10USD_DAY",
+                price = Price(amount = 10_00, currency = "USD"),
+                properties = mapOf(
+                        PRODUCT_CLASS.s to SUBSCRIPTION.name,
```
Similar to another comment, product class should be MEMBERSHIP.
Payment type is SUBSCRIPTION.
If no payment type is set, then the default value is ONE TIME PAYMENT.

vihangpatil commented on neo4j-store/src/main/resources/AcceptanceTestSetup.kts:
```diff
@@ -126,4 +129,71 @@ adminStore.atomicCreateOffer(
+                properties = mapOf(
+                        PRODUCT_CLASS.s to SUBSCRIPTION.name,
+                        SEGMENT_IDS.s to "country-us"),
+                payment = mapOf(
```
The payment type = subscription has to be set in payment properties.

vihangpatil commented on acceptance-tests/src/main/kotlin/org/ostelco/at/common/Products.kt:
```diff
@@ -32,6 +32,24 @@ val expectedPlanProduct: Product = Product()
+                        "label" to "Daily subscription plan",
+                        "taxRegionId" to "us"
+                )
+        )
```
Here, the payment type is set correctly to "SUBSCRIPTION", but the product class should be MEMBERSHIP.

vihangpatil commented on customer-endpoint/src/main/kotlin/org/ostelco/prime/customer/endpoint/resources/ProductsResource.kt:
```diff
@@ -54,4 +55,31 @@ class ProductsResource(private val dao: SubscriberDAO) {
+    @POST
+    @Path("{sku}/renew")
+    @Produces(MediaType.APPLICATION_JSON)
+    fun renewSubscription(@Auth token: AccessTokenPrincipal?,
```
Just the word `Subscription` means mobile Subscription (with MSISDN as ID).  Better to call this `PaymentSubscription` or `PlanSubscription` to distinguish it from "Mobile" Subscription.

vihangpatil commented on app-notifier/src/main/kotlin/org/ostelco/prime/appnotifier/FirebaseAppNotifier.kt:
```diff
@@ -45,6 +45,24 @@ class FirebaseAppNotifier: AppNotifier {
+                    /* No notification, as a notification will be sent on either successful
+                       or failed renewal. */
+                }
             }
```
This API is a bit messed up.  `enum NotificationType` can have fields such as `title` and `body` which is then directly used instead of using `when` cases.  Also, if FCMStrings is specific to Firebase, they should not be in generic API interfaces - `prime-modules`.

The simplest way to design it is to think of `AppNotifier` as an interface whereas `FirebaseAppNotifier` as one of the many possible implementations of `AppNotifier`.
We should be able to replace whole FCM with some other implementation with just adding and removing dependency in the gradle build file and no need of any code change.

kmmatwork replied:
https://www.notion.so/redotter/Review-FCM-delivery-setup-reliability-and-message-format-0701c7da03e24a89963cc7bdd4f7aa76

vihangpatil commented on payment-processor/src/main/kotlin/org/ostelco/prime/paymentprocessor/StripePaymentProcessor.kt:
```diff
@@ -458,10 +513,34 @@ class StripePaymentProcessor : PaymentProcessor {
+                        *(if (sourceId != null)
+                            arrayOf("source" to sourceId)
+                        else arrayOf())
+                )
```
Outer if-else with mapOf inside is more readable.
Also, ?. is more compact way to check nulls.

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Renames data class 'SubscriptionStateInfo' to 'SubscriptionPaymentInfo'

Commit:
Moves 'subscription' tag from product 'product-class' to product 'payment-type'

Commit:
Minor fixes and clarifications

kmmatwork commented:
Fyi, more changes to how plans/products gets stored will follow later as part of the 
https://www.notion.so/redotter/Document-how-Backend-should-handle-prorate-price-plan-changes-for-a-given-Product-39d34912997c4a74870b1960c9fa08cb
and the 
https://www.notion.so/redotter/Change-implementation-of-Purchase-Records-to-store-Price-information-from-Stripe-Invoices-f3ac8e7f00c84cc1b60d01ef5512712a
tasks.

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Renames functions related to payment subscriptions in order to
distinguish them from functions related to mobile subscriptions

Commit:
Switch from ngrok.io to serveo.net for supporting webhook in acceptance tests

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge remote-tracking branch 'origin/feature/payment-recurring' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Make check of STRIPE_ENDPOINT_SECRET variable more explisit in conditional acceptance test suite

Commit:
Switch from serveo.net to stripe-cli Docker image for supporting webhook in acceptance tests

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Adds documentation for test setup ot the recurring payment acceptance tests

Commit:
Disable the recurring payment acceptance tests

The @EnabledIfEnvironmentVariable annotation is not working as
expected...

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Adds fixes for using Stripe CLI as proxy for Stripe events

Changes:
  - Don't start Stripe CLI as proxy unless the STRIPE_ENDPOINT_SECRET
    env. variable is not set or set to a dummy value.
  - The @EnabledIfEnvironmentVariable Junit annontation seems not
    to be working. Adds an additional check to ensure that the
    recurring payment tests are not started when the
    STRIPE_ENDPOINT_SECRET is not set or set to a dummy value.

Commit:
Adds missing Stripe proxy Dockerfile files

Commit:
Adds fixes to test setup using Stripe proxy

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

la3lma commented on acceptance-tests/script/wait.sh:
```diff
@@ -220,7 +220,18 @@ EOF
+##
+
+echo "Creating pub-sub subscriptions for recurring payment tests"
+curl -X PUT -H "Content-Type: application/json" -d '{"topic":"projects/'${GCP_PROJECT_ID}'/topics/stripe-event","ackDeadlineSeconds":10}' pubsub-emulator:8085/v1/projects/${GCP_PROJECT_ID}/subscriptions/stripe-event-okhttp-purchase-ok-sub
```
This is so much repeated code that it now arguably would make sense to factor this out into a function.   Or not. Your call.

kmmatwork replied:
Will try to move the `curl` call into a function.

la3lma commented on acceptance-tests/src/main/kotlin/org/ostelco/at/common/StripeEventListener.kt:
```diff
@@ -0,0 +1,167 @@
+                                         subscription: String,
+                                         customerId: String,
+                                         timeout: Long = 10000L): Boolean =
+            waitForStripeEvent(
```
This code  has very convoluted control flow. Consider breaking it up into smaller pieces with descriptive names.

kmmatwork replied:
Will try to move the `cmp` function out out.

la3lma commented on acceptance-tests/src/main/kotlin/org/ostelco/at/common/StripeEventListener.kt:
```diff
@@ -0,0 +1,167 @@
+                        delay(200L)
+                    }
+                    "Done"
+                } != null
```
This makes it very obscure what we're checking for.   Please consider making a local function with a descriptive name.  Code should preferably be trivial to read, and this isn't.

kmmatwork replied:
Basically cut and paste from [https://kotlinlang.org/docs/reference/coroutines/cancellation-and-timeouts.html](https://kotlinlang.org/docs/reference/coroutines/cancellation-and-timeouts.html) at the bottom of the page.

la3lma commented on acceptance-tests/src/main/kotlin/org/ostelco/at/common/StripePayment.kt:
```diff
@@ -66,6 +69,29 @@ object StripePayment {
+                        ),
+                        "email" to "me@somewhere.com")
+        )
+        val source = Source.create(sourceMap)
```
I would have preferred calling this paymentSource, just to make it blindingly obvious to casual readers (such as myself, in the role of reviewer) what this value is.

kmmatwork replied:
Good point!

la3lma commented on acceptance-tests/src/main/kotlin/org/ostelco/at/jersey/Tests.kt:
```diff
@@ -1897,7 +1899,157 @@ class PlanTest {
+                queryParams = mapOf("sourceId" to sourceId)
+            }
+
+            Thread.sleep(200) // wait for 200 ms for balance to be updated in db
```
This is brittle.   Is there a less brittle way to wait for the database?

kmmatwork replied:
Not sure. The same `wait` statement is present in all acceptance tests that involves updates to `neo4j`. Will add a TODO and a Notion tech dept task.

la3lma commented on acceptance-tests/src/main/kotlin/org/ostelco/at/jersey/Tests.kt:
```diff
@@ -1897,7 +1899,157 @@ class PlanTest {
+            // expires, which will happen after 4 sec. Waiting a bit longer
+            // before checking the outcome.
+            StripeEventListener.waitForSubscriptionPaymentToSucceed(
+                    subscription = "stripe-event-jersey-purchase-ok-sub",
```
What does this magic string mean? Consider putting it into a variable with a descriptive name.

kmmatwork replied:
Name of a  subscription to the "stripe-event" pub-sub topic. We need one such subscription for each "recurring payment" acceptance test. The  `waitForSubscriptionPaymentToSucceed` method is there to ensure that we don't wait too long for the expected events.

Not sure if it make sense use a variable for the subscription name. But some more comments about what is going on could be an idea.

la3lma commented on acceptance-tests/src/main/kotlin/org/ostelco/at/okhttp/Tests.kt:
```diff
@@ -994,7 +996,125 @@ class PlanTest {
+            customerId = createCustomer(name = "Test Purchase Plan User", email = email).id
+            enableRegion(email = email, region = "us")
+
+            val client = clientForSubject(subject = email)
```
There is a lot of repeated code between this and the previous test.   Consider factoring that code into separate methods or perhaps even running the code in a a "setUp" method as part off fixture setup for the test suite.

Repeated code obscures the intent and implementation of the actual test.

kmmatwork replied:
Will try to restructure a bit.

la3lma commented on acceptance-tests/src/main/kotlin/org/ostelco/at/okhttp/Tests.kt:
```diff
@@ -994,7 +996,125 @@ class PlanTest {
+            // Actual charge for renewal will first be done after trial time
+            // expires, which will happen after 4 sec. Waiting a bit longer
+            // before checking the outcome.
+            StripeEventListener.waitForFailedSubscriptionRenewal(
```
Repeated code

la3lma commented on acceptance-tests/src/main/kotlin/org/ostelco/at/okhttp/Tests.kt:
```diff
@@ -994,7 +996,125 @@ class PlanTest {
+            // Explicitly renew the subscription with a new card.
+            client.renewSubscription(sku, newSourceId, true)
+
+            StripeEventListener.waitForSubscriptionPaymentToSucceed(
```
Repeated code

la3lma commented on customer-endpoint/src/main/kotlin/org/ostelco/prime/customer/endpoint/resources/ProductsResource.kt:
```diff
@@ -54,4 +54,31 @@ class ProductsResource(private val dao: SubscriberDAO) {
+                                 sourceId: String?,
+                                 @QueryParam("saveCard")
+                                 saveCard: Boolean?): Response =
+            if (token == null) {
```
Can authorisation check be delegated to a filter so that it does not show up in the business logic part of the resource implementation?  Technically it obviously can, but the question is obviously if it makes sense in this particular situation.  If  the "token == null"  must be there, add a comment describing why this couldn't be delegated to a filter, of if it can, at the very least add a "TODO" indicating that it should.

kmmatwork replied:
Agree with this. I think the same pattern is used in all client facing APIs, and an update should probably fix it for all of them. Will leave a TODO and a Notion tech dept task for fixing this.

la3lma commented on customer-endpoint/src/main/kotlin/org/ostelco/prime/customer/endpoint/resources/ProductsResource.kt:
```diff
@@ -54,4 +54,31 @@ class ProductsResource(private val dao: SubscriberDAO) {
+                if (sourceId == null) {
+                    dao.renewPaymentSubscription(
+                            identity = token.identity,
+                            sku = sku)
```
In this case the saveCard parameter is discarded?   If it's non-null that should  indicate an error, or it should be used.

kmmatwork replied:
Will fix.

la3lma commented on customer-endpoint/src/main/kotlin/org/ostelco/prime/customer/endpoint/resources/ProductsResource.kt:
```diff
@@ -54,4 +54,31 @@ class ProductsResource(private val dao: SubscriberDAO) {
+                    dao.renewPaymentSubscription(
+                            identity = token.identity,
+                            sku = sku,
+                            sourceId = sourceId,
```
What is the default value of sourceId? If it's null, then perhaps only the "else" branch is necessary and the code can be made more compact without losing generality?

kmmatwork replied:
Yes, this API needs some rethinking.

la3lma commented on acceptance-tests/src/main/kotlin/org/ostelco/at/common/StripeEventListener.kt:
```diff
@@ -0,0 +1,167 @@
+                    while (!subscriber.eventSeen()) {
+                        delay(200L)
+                    }
+                    "Done"
```
This control flow is very obscure. Can it be made more obvious?

la3lma commented on acceptance-tests/src/main/kotlin/org/ostelco/at/common/StripeEventListener.kt:
```diff
@@ -0,0 +1,167 @@
+                status = withTimeoutOrNull(timeout) {
+                    subscriber.start()
+                    while (!subscriber.eventSeen()) {
+                        delay(200L)
```
Why the delay, why 200, which unit is it (milliseconds).  Please explain, at least in a comment.

la3lma commented on acceptance-tests/src/main/kotlin/org/ostelco/at/common/StripeEventListener.kt:
```diff
@@ -0,0 +1,167 @@
+                    eventType = "customer.subscription.updated",
+                    timeout = timeout,
+                    cmp = {
+                        if (it.type != "customer.subscription.updated")
```
Obscure controll flow. Make it more obvious what's going on.

la3lma commented on acceptance-tests/src/main/kotlin/org/ostelco/at/common/StripeEventListener.kt:
```diff
@@ -0,0 +1,167 @@
+        try {
+            gson.fromJson(message.toStringUtf8(), Event::class.java)
+                    .let {
+                        if (cmp(it)) eventSeen = true
```
What does this mean? My initial reading is  "compare it with itself", which makes no sense.  A comment is in order.

la3lma commented on acceptance-tests/src/main/kotlin/org/ostelco/at/common/StripePayment.kt:
```diff
@@ -105,6 +131,7 @@ object StripePayment {
     fun getStripeCustomerId(customerId: String) : String {
+
         // https://stripe.com/docs/api/java#create_card_token
         Stripe.apiKey = System.getenv("STRIPE_API_KEY")
```
This thing doesn't test if the STRIPE_API_KEY is non-null. It's clearly an error if it is, and that error should be detected and signalled.

la3lma commented on acceptance-tests/src/main/kotlin/org/ostelco/at/common/StripePayment.kt:
```diff
@@ -121,16 +148,21 @@ object StripePayment {
     fun deleteCustomer(customerId: String) {
+
         // https://stripe.com/docs/api/java#create_card_token
         Stripe.apiKey = System.getenv("STRIPE_API_KEY")
```
Getting the stripe API key should be wrapped in a method that checks if it has a valid value (at least that it isn't empty), and throws/signals the appropriate error if it's not valid.   The rationale behind this request is: This will fail eventually, so it is prudent to fail as early as possible to make sure that half-committed values are not stored anywhere before the error is detected.

la3lma commented on acceptance-tests/src/main/kotlin/org/ostelco/at/jersey/Tests.kt:
```diff
@@ -1897,7 +1899,157 @@ class PlanTest {
+            // expires, which will happen after 4 sec. Waiting a bit longer
+            // before checking the outcome.
+            StripeEventListener.waitForFailedSubscriptionRenewal(
+                    subscription = "stripe-event-jersey-purchase-fail-sub",
```
This (along with other things) in this test is repeated code. Repeated code makes the code less readable in the sense that it's difficult to know what's special in something that looks similar but can't be known to be similar before a complete inspection has been done.  It is better to factor this out into a separate variable or method (depending on the context) so that it is obvious, at a glance, what is identical and what isn't, and also what the differences are (can be highlighted using parameters to methods)).

la3lma commented on acceptance-tests/src/main/kotlin/org/ostelco/at/okhttp/Tests.kt:
```diff
@@ -994,7 +996,125 @@ class PlanTest {
+            // Actual charge for renewal will first be done after trial time
+            // expires, which will happen after 4 sec. Waiting a bit longer
+            // before checking the outcome.
+            StripeEventListener.waitForSubscriptionPaymentToSucceed(
```
Repeated code

la3lma commented on acceptance-tests/src/main/kotlin/org/ostelco/at/okhttp/Tests.kt:
```diff
@@ -994,7 +996,125 @@ class PlanTest {
+                    timeout = 30000L)
+
+            // Wait for DB to be updated.
+            Thread.sleep(200)
```
Sleeping is a very brittle way to ensure that a database is updated. Is there a better way?

la3lma commented on acceptance-tests/src/main/kotlin/org/ostelco/at/okhttp/Tests.kt:
```diff
@@ -994,7 +996,125 @@ class PlanTest {
+            // expires, which will happen after 4 sec. Waiting a bit longer
+            // before checking the outcome.
+            StripeEventListener.waitForFailedSubscriptionRenewal(
+                    subscription = "stripe-event-okhttp-purchase-fail-sub",
```
Repeated code

la3lma commented on model/src/main/kotlin/org/ostelco/prime/model/Entities.kt:
```diff
@@ -332,6 +337,9 @@ data class RefundRecord(
         val reason: String, // possible values are duplicate, fraudulent, and requested_by_customer
         val timestamp: Long) : HasId
 
+/* TODO! (kmm) To support prorating with subscriptions, the
```
what is "prorating"?

kmmatwork replied:
Standard term for ensuring correct charges when changing/terminating subscriptions etc. More info at [https://stripe.com/docs/billing/subscriptions/prorations](https://stripe.com/docs/billing/subscriptions/prorations) for how to use it with Stripe. Also described many other places.

la3lma commented on neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt:
```diff
@@ -1163,18 +1173,29 @@ object Neo4jStoreSingleton : GraphStore {
-                    val (chargeId, invoiceId) = when (product.paymentType) {
-                        SUBSCRIPTION -> {
-                            val subscriptionDetailsInfo = purchasePlan(
+                    val purchaseRecord = when (product.paymentType) {
```
This is a very long if-branch.  Consider making a method with a descriptive name (a scope-local one if convenient). This will increase the readability of the code.

la3lma commented on neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt:
```diff
@@ -2437,45 +2468,221 @@ object Neo4jStoreSingleton : GraphStore {
+                                         sku: String,
+                                         sourceId: String,
+                                         saveCard: Boolean): Either<StoreError, Product> = writeTransaction {
+        getCustomer(identity = identity)
```
This is highly unreadable code.  The control flow is obscure, the level of indentation is high.  Please introduce other abstractions (local methods probably) to make the control flow and thus semantics of this statement comprehensible.  Right now it isn't.

la3lma commented on neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt:
```diff
@@ -2437,45 +2468,221 @@ object Neo4jStoreSingleton : GraphStore {
+                                                         invoiceId: String,
+                                                         payInvoiceNow: Boolean): Either<StoreError, Unit> = readTransaction {
+        get(Product withSku sku)
+                .flatMap { product ->
```
Obscure control flow. See comment above about similar issue.

la3lma commented on neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt:
```diff
@@ -2437,45 +2468,221 @@ object Neo4jStoreSingleton : GraphStore {
+                        "for renewal of subscription to plan ${plan.id}"
+                )
                 plan
             }.fix()
```
This boilerplate adds two levels of indentation and obscures control flow.  Can it be abstracted away into something more compact?

la3lma commented on neo4j-store/src/main/kotlin/org/ostelco/prime/storage/graph/Neo4jStore.kt:
```diff
@@ -2437,45 +2468,221 @@ object Neo4jStoreSingleton : GraphStore {
+                        ).plus(product.presentation))
+                Unit
+            }.fix()
+        }.unsafeRunSync()
```
Same as above, can some abstraction with a lambda ba used instead, e.g. "withinTransaction(transaction) { ...}" or something?

la3lma commented on payment-processor/src/main/kotlin/org/ostelco/prime/paymentprocessor/StripePaymentProcessor.kt:
```diff
@@ -196,6 +209,35 @@ class StripePaymentProcessor : PaymentProcessor {
+    override fun addSource(customerId: String,
+                           sourceId: String,
+                           setDefault: Boolean): Either<PaymentError, SourceInfo> =
+            either("Failed to add source $sourceId to customer $customerId") {
```
Very obscure control flow.  Please rewrite for readability (whole method)

kmmatwork replied:
Have to disagree here, provided that you mean the `either()` call. The `either` wrapper i a `try catch` wrapper that will turn exception thrown on calls to Stripe into an `Either`. Used in all methods calling Stripe.

la3lma commented on payment-processor/src/main/kotlin/org/ostelco/prime/paymentprocessor/StripePaymentProcessor.kt:
```diff
@@ -223,63 +265,75 @@ class StripePaymentProcessor : PaymentProcessor {
                         "items" to mapOf("0" to item),
-                        *(if (trialEnd > Instant.now().epochSecond)
-                            arrayOf("trial_end" to trialEnd.toString())
+                        *(if (trialEndInEpochSeconds > Instant.now().epochSecond)
```
What does the asterisk mean in this context? This is necessarily obscure.

kmmatwork replied:
Kotlins [spread](https://kotlinlang.org/docs/reference/functions.html#variable-number-of-arguments-varargs) operator.

la3lma commented on payment-processor/src/main/kotlin/org/ostelco/prime/paymentprocessor/StripePaymentProcessor.kt:
```diff
@@ -223,63 +265,75 @@ class StripePaymentProcessor : PaymentProcessor {
+            }
+            when (subscription.status) {
+                "active" -> {
+                    when (intent.status) {
```
Please avoid using magic strings like "succeed", "incomplete" etc. If these refer to protocol options, then please use enumerations with strings in them to identify the different protocol options.   Naked (and thus magic) strings obscure what's going on, and are also difficult to keep track of if used more than one place in the codebase (and it's hard to know if it's used more than once since text search is necessary, it's much easier to  jump to the enumeration and then see where it's used).

la3lma commented on payment-processor/src/main/kotlin/org/ostelco/prime/paymentprocessor/StripePaymentProcessor.kt:
```diff
@@ -223,63 +265,75 @@ class StripePaymentProcessor : PaymentProcessor {
+        }.let {
+            SubscriptionPaymentInfo(id = subscription.id,
+                    status = it,
+                    invoiceId = if (invoice != null)
```
consider using either braces in branches of if/else or wrapping it in a () (or both).

la3lma commented on payment-processor/src/main/kotlin/org/ostelco/prime/paymentprocessor/StripeUtils.kt:
```diff
@@ -23,29 +24,33 @@ object StripeUtils {
+            /* Something is wrong with the payment source/card.
+               402 indicates that the card is out of funds or expired etc. */
+            logger.warn("Payment error: ${errorDescription}, Stripe Error Code: ${e.code}")
+            Either.left(if (e.statusCode == 402)
```
Please don't use naked numbers.   Find a constant that represents the semantics of this particular return code (I didn't remember  off hand that 402 means "payment required", something that would have been obvious if a descriptive enumeration name had been used instead). There are plenty of them in the various HTTP utility classes we use, please use one of them.

la3lma commented on payment-processor/src/main/kotlin/org/ostelco/prime/paymentprocessor/StripeUtils.kt:
```diff
@@ -23,29 +24,33 @@ object StripeUtils {
+            Either.left(if (e.statusCode == 402)
+                PaymentFailedError(errorDescription, e.message)
+            else
+                ForbiddenError(errorDescription, e.message))
```
What is forbidden error? An error because something is forbidden, or an error that should never happen?   If it's the latter then the method should be given a better name,, otherwise some comment should indicate why the invocation was forbidden.

la3lma commented on payment-processor/src/main/kotlin/org/ostelco/prime/paymentprocessor/subscribers/RecurringPaymentStripeEvent.kt:
```diff
@@ -49,57 +58,146 @@ class RecurringPaymentStripeEvent : PubSubSubscriber(
-        /* Skip invoices not related to subscriptions (recurring payment). */
         if (invoice.subscription.isNullOrEmpty())
             return
+        if (invoice.billingReason != "subscription_cycle")
```
Magic string! Use string constant or string associated with enumeration instead.

la3lma commented on payment-processor/src/main/kotlin/org/ostelco/prime/paymentprocessor/subscribers/RecurringPaymentStripeEvent.kt:
```diff
@@ -49,57 +58,146 @@ class RecurringPaymentStripeEvent : PubSubSubscriber(
-            "invoice.payment_failed" -> {
-            }
+            /* Early notification (configurable in Stripe console) of renewal. */
             "invoice.upcoming" -> {
```
More magic strings.

la3lma commented on prime-modules/src/main/kotlin/org/ostelco/prime/apierror/ApiErrorCodes.kt:
```diff
@@ -65,6 +65,7 @@ enum class ApiErrorCode {
     FAILED_TO_REMOVE_PLAN,
     ALREADY_SUBSCRIBED_TO_PLAN,
+    FAILED_TO_RENEW_SUBSCRIPTION,
     // FAILED_TO_SUBSCRIBE_TO_PLAN,
```
If this shouldn't be enabled in the immediate  future, then remove the dead code.

la3lma commented on prime-modules/src/main/kotlin/org/ostelco/prime/paymentprocessor/core/Model.kt:
```diff
@@ -18,18 +18,37 @@ data class ProfileInfo(val id: String)
+                                   val invoiceId: String,           /* Not set on plans with trial time. */
+                                   val chargeId: String? = null,    /* Only set on successful payment on creation/renewal. */
+                                   val created: Long,
+                                   val currentPeriodStart: Long,
```
Add a comment indicating what the times should be interpreted as? Is it seconds since epoch? If so, say so.

kmmatwork replied:
Internally timestamps are always in milliseconds since epoch.

la3lma commented on prime/infra/dev/prime-customer-api.yaml:
```diff
@@ -651,6 +651,39 @@ paths:
+          description: "Successfully renewed the subscription"
+          schema:
+            $ref: '#/definitions/Product'
+        404:
```
Wasn't there a 402 in there somewhere also?

la3lma commented on payment-processor/src/main/kotlin/org/ostelco/prime/paymentprocessor/StripePaymentProcessor.kt:
```diff
@@ -458,10 +513,33 @@ class StripePaymentProcessor : PaymentProcessor {
             either("Failed to complete payment of invoice ${invoiceId}") {
-                val receipt = Invoice.retrieve(invoiceId).pay()
-                InvoicePaymentInfo(receipt.id, receipt?.charge ?: UUID.randomUUID().toString())
+                val params = if (sourceId != null)
```
Please wrap if in parens.  I get confused by this kind of implicitly bounded values.  Be explicit.

kmmatwork replied:
Agree with that.

la3lma commented on acceptance-tests/src/main/kotlin/org/ostelco/at/common/StripeEventListener.kt:
```diff
@@ -0,0 +1,167 @@
+
+        try {
+            runBlocking {
+                status = withTimeoutOrNull(timeout) {
```
If this is just a busy-wait loop, then say so in a comment. If not then explain what's going on.

la3lma commented on acceptance-tests/src/main/kotlin/org/ostelco/at/common/StripeEventListener.kt:
```diff
@@ -0,0 +1,167 @@
+                        }
+                    })
+
+    fun waitForFailedSubscriptionRenewal(topic: String = StripeEventListener.topic,
```
The control flow of this method is very obscure. Please make it simpler to read by breaking it up into smaller methods with descriptive names rather than one big open-coded method with a  very complex internal life.

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Fixes overlooked incorrect class name from last manual merge with develop using github UI

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Fixes after last merge with develop

Commit:
Adds 'us' as the third region to acceptance tests as it is used
in the recurring payment tests

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring

Commit:
Merge branch 'develop' into feature/payment-recurring
